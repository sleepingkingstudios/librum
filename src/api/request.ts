import type {
  Annotated,
  Annotations,
} from '@utils/annotations';
import { applyMiddleware } from '@utils/middleware';
import type {
  Middleware,
  MiddlewareBuilder,
} from '@utils/middleware';
import type { DataObject } from '@utils/types';
import type { Mutation } from './types';

export type Request = (
  (values?: DataObject) => Promise<unknown> | void
) & Annotated;

type UseRequestParams<Options extends DataObject = DataObject> = {
  mutation: Mutation,
  options?: Options,
};

type BuildRequestParams = {
  middleware?: MiddlewareBuilder[] | MiddlewareBuilder,
  params?: DataObject,
};

export type UseRequest<Options extends DataObject = DataObject> = (
  ({
    mutation,
    options,
  }: UseRequestParams<Options>) => Request
) & Annotated;

export type BuildRequest<Options extends DataObject = DataObject> = (
  (
    {
      middleware,
      params,
    }: BuildRequestParams,
    annotations?: Annotations,
  ) => UseRequest<Options>
) & Annotated;

export const buildRequest: BuildRequest = <Options extends DataObject = DataObject>(
  {
    middleware = [] as MiddlewareBuilder[],
    params = {},
  }: BuildRequestParams,
  annotations: Annotations = null,
) => {
  const middlewareArray: MiddlewareBuilder[] =
    Array.isArray(middleware) ? middleware : [middleware];
  const useRequest: UseRequest<Options> = ({
    mutation,
    options = {},
  }) => {
    const reifiedMiddleware: Middleware[] =
      middlewareArray.map(builder => builder(options));

    const request: Request = async (values = {}) => {
      return await applyMiddleware(mutation, reifiedMiddleware)({
        ...params,
        ...values,
      });
    };

    if (annotations !== null) {
      request.annotations = {
        middleware: reifiedMiddleware,
        mutation,
        options,
        params,
        type: 'request',
        ...annotations,
      };
    }

    return request;
  };

  if (annotations !== null) {
    useRequest.annotations = {
      middleware: middlewareArray,
      params,
      type: 'hooks:useRequest',
      ...annotations,
    };
  }

  return useRequest;
};
