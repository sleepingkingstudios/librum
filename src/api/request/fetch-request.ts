import type {
  FetchOptions,
  HttpMethod,
  PerformRequest,
  RequestOptions,
  RequestParams,
  RequestWildcards,
  Response,
  ResponseData,
} from '../types';
import {
  applyWildcards,
  formatBody,
  formatParams,
  mapData,
  withData,
  withError,
  withStatus,
} from '../utils';

type DefaultsWithOptions = {
  contentType?: string,
  fetchOptions: FetchOptions,
  params?: RequestParams,
  wildcards?: RequestWildcards,
};

const applyDefaults = ({
  contentType = 'json',
  headers = {},
  ...options
}: RequestOptions): DefaultsWithOptions => {
  const {
    body,
    method,
    params,
    wildcards,
    ...rest
  } = options;
  const formattedBody = formatBody({ body });
  const fetchOptions: FetchOptions = {
    ...rest,
    ...(formattedBody ? { body: formattedBody } : {}),
    headers: {
      ...defaultHeaders({ contentType }),
      ...headers,
    },
    method: (method || 'get').toUpperCase() as HttpMethod,
  };

  return {
    contentType,
    fetchOptions,
    params,
    wildcards,
  };
};

const defaultHeaders = ({
  contentType,
}: {
  contentType: string,
}): Record<string, string> => {
  const headers: Record<string, string> = {}

  if (contentType.toLowerCase() === 'json') {
    headers['Content-Type'] = 'application/json';
  } else {
    headers['Content-Type'] = 'text/plain';
  }

  return headers;
};

const isError = (error: unknown): boolean => {
  if (!(typeof error === 'object')) { return false; }

  return 'message' in error && 'name' in error;
};

export const fetchRequest: PerformRequest = (url, options = {}) => {
  const {
    contentType,
    fetchOptions,
    params,
    wildcards,
  } = applyDefaults(options);

  let ok: boolean;

  try {
    const fetchUrl =
      `${applyWildcards({ url, wildcards })}${formatParams(params)}`;

    return fetch(fetchUrl, fetchOptions)
      .then((res) => {
        ok = res.ok;

        if (contentType.toLowerCase() === 'json') {
          return res.json();
        } else {
          return res.text();
        }
      })
      .then((raw) => {
        const body = raw as ResponseData;
        const data = mapData(body);

        let response: Response = withStatus({
          status: (ok ? 'success' : 'failure'),
        });

        if (data) { response = withData({ data, response }); }

        return new Promise(resolve => resolve(response));
      })
      .catch((error: unknown) => {
        if (!isError(error)) { throw error; }

        const { name } = error as Record<string, unknown>;

        if (name !== 'SyntaxError') { throw error; }

        const response = withError({
          error: error.toString(),
          response: withStatus({ status: 'errored' }),
        });

        return new Promise(resolve => resolve(response));
      }) as Promise<Response>;
  } catch(error: unknown) {
    const response = withError({
      error: error.toString(),
      response: withStatus({ status: 'errored' }),
    });

    return new Promise(resolve => resolve(response));
  }
};
