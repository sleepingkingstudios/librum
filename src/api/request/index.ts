export { fetchRequest } from './fetch-request';
