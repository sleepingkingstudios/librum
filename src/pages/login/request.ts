import { faUserSlash } from '@fortawesome/free-solid-svg-icons';

import { useApiRequest } from '@api';
import type {
  AlertDirective,
  Middleware,
  MiddlewareOptions,
  PerformRequest,
  RequestOptions,
  Response,
  ResponseData,
} from '@api';
import type {
  Session,
  User,
} from '@session';
import { actions as sessionActions } from '@session/reducer';
import {
  clearStoredSession,
  setStoredSession,
} from '@session/utils';
import type {
  Action,
  Dispatch,
} from '@store';

const buildSession = (data: ResponseData): Session => {
  if (data === null || data === undefined) { return { authenticated: false }; }

  if (typeof data === 'string') { return { authenticated: false }; }

  if (!('token' in data && 'user' in data)) { return { authenticated: false }; }

  const token = data.token as string;
  const user = data.user as User;

  return {
    authenticated: true,
    token,
    user,
  };
};
const loginAlerts: AlertDirective[] = [
  {
    dismiss: 'authentication:session',
    status: 'success',
  },
  {
    display:  {
      context: 'authentication:session',
      icon: faUserSlash,
      message: 'User not found with the provided username and password.',
      type: 'failure',
    },
    status: 'failure',
  }
];

export const createSessionMiddleware: Middleware =
  (fn: PerformRequest, config: MiddlewareOptions): PerformRequest => {
    const dispatch = config.dispatch as Dispatch;
    const { create, destroy } = sessionActions;

    return async (url: string, options?: RequestOptions): Promise<Response> => {
      const response = await fn(url, options);
      const { data, isSuccess } = response;

      if (!isSuccess) { return response; }

      const session: Session = buildSession(data);
      const { authenticated, token, user } = session;

      if (authenticated) {
        const action: Action<Session> = create({ token, user });

        dispatch(action);

        setStoredSession(session);
      } else {
        const action: Action<Session> = destroy();

        dispatch(action);

        clearStoredSession();
      }

      return response;
    };
  };

export const useLoginRequest = () => {
  return useApiRequest({
    alerts: loginAlerts,
    method: 'post',
    middleware: [
      createSessionMiddleware,
    ],
    url: '/api/authentication/session',
  });
};
