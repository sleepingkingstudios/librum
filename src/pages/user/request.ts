import { faUser } from '@fortawesome/free-solid-svg-icons';

import { useApiQuery } from '@api';
import type { AlertDirective } from '@api';

const alerts: AlertDirective[] = [
  {
    status: 'failure',
    display: {
      context: 'pages:user:request',
      icon: faUser,
      message: 'Unable to load current user.',
      type: 'failure',
    },
  },
  {
    status: 'success',
    dismiss: 'pages:user:request',
  },
];

export const useGetUserQuery = () => useApiQuery({
  alerts,
  url: '/api/authentication/user',
});
