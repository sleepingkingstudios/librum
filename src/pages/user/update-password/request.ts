import {
  faUserLock,
  faUserSlash,
} from '@fortawesome/free-solid-svg-icons';

import { useApiRequest } from '@api';
import type {
  AlertDirective,
  Middleware,
  MiddlewareOptions,
  PerformRequest,
  RequestOptions,
  Response,
} from '@api';
import { invalidPasswordError } from '@api/errors';

type CloseForm = () => void;

const alerts: AlertDirective[] = [
  {
    errorType: invalidPasswordError,
    display: {
      context: 'pages:user:updatePassword:alerts',
      icon: faUserSlash,
      message: 'Password does not match current password.',
      type: 'failure',
    },
  },
  {
    status: 'failure',
    display: {
      context: 'pages:user:updatePassword:alerts',
      icon: faUserSlash,
      message: 'Unable to update password.',
      type: 'failure',
    },
  },
  {
    status: 'success',
    display: {
      context: 'pages:user:updatePassword:alerts',
      icon: faUserLock,
      message: 'Successfully updated password.',
      type: 'success',
    },
  },
];

export const closeFormMiddleware: Middleware =
  (fn: PerformRequest, config: MiddlewareOptions) => {
    const closeForm = config.closeForm as CloseForm;

    return async (url: string, options?: RequestOptions): Promise<Response> => {
      const response: Response = await fn(url, options);
      const { isSuccess } = response;

      if (isSuccess) { closeForm(); }

      return response;
    };
  };

export const useUpdateUserPasswordRequest = ({
  config,
}: {
  config: MiddlewareOptions,
}) => useApiRequest({
  alerts,
  config,
  method: 'patch',
  middleware: [closeFormMiddleware],
  url: '/api/authentication/user/password',
});
