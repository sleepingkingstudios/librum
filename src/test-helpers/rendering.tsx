import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
