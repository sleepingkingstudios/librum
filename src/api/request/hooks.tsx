import * as React from 'react';

import { fetchRequest } from './fetch-request';
import type {
  Refetch,
  RefetchOptions,
  RequestOptions,
  Response,
} from './types';
import { withStatus } from './utils';

export const useRequest = ({
  options = {},
  url,
}: {
  options?: RequestOptions,
  url: string,
}): [Response, Refetch] => {
  const initialStatus = 'uninitialized';
  const [response, setResponse] =
    React.useState<Response>(withStatus({ status: initialStatus }));
  const {
    body,
    headers,
    method,
    params,
    wildcards,
  } = options;

  const performRequest: Refetch = (request?: RefetchOptions): void => {
    const { status } = response;

    if (status === 'loading') { return; }

    setResponse(withStatus({ response, status: 'loading' }));

    fetchRequest(
      url,
      {
        body,
        headers,
        method,
        params,
        wildcards,
        ...request,
      }
    )
      .then(response => setResponse(response))
      .catch(error => { throw error; });
  };

  return [response, performRequest];
};
