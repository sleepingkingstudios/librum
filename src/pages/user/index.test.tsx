import * as React from 'react';

import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { UserPage } from './index';
import {
  responseWithData,
  responseWithStatus,
} from '@api';
import type { Response } from '@api';
import type { User } from '@session';
import { render } from '@test-helpers/rendering';
import { useGetUserQuery } from './request';

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@components/page', () => require('@components/page/mocks'));
// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@core/navigation', () => require('@core/navigation/mocks'));
jest.mock('./request');

const mockUseQuery =
  useGetUserQuery as jest.MockedFunction<typeof useGetUserQuery>;

describe('<UserPage>', () => {
  const refetch = jest.fn();

  describe('when the user is loading', () => {
    const response: Response = responseWithStatus({ status: 'loading' });

    beforeEach(() => {
      mockUseQuery.mockImplementation(() => [response, refetch]);
    });

    it('should display the loading overlay', () => {
      const { getByText } = render(
        <UserPage />,
        { store: true },
      );

      const overlay = getByText('Loading User...');

      expect(overlay).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <UserPage />,
        { store: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the user has failed to load', () => {
    const response: Response = responseWithStatus({ status: 'failure' });

    beforeEach(() => {
      mockUseQuery.mockImplementation(() => [response, refetch]);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <UserPage />,
        { store: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the user has successfully loaded', () => {
    const user: User = {
      id: '0',
      email: 'alan.bradley@example.com',
      role: 'user',
      slug: 'alan-bradley',
      username: 'Alan Bradley',
    };
    const response: Response = responseWithData({
      data: { user },
    });

    beforeEach(() => {
      mockUseQuery.mockImplementation(() => [response, refetch]);
    });

    it('should render the user information', () => {
      const expectedLabels = [
        'Name',
        'Email',
        'Role',
      ];
      const expectedDefinitions = [
        user.username,
        user.email,
        'User',
      ];

      const { queryAllByRole } = render(
        <UserPage />,
        { store: true },
      );

      const labels = queryAllByRole('term');

      expect(labels.map(
        (label) => label.textContent)
      ).toEqual(expectedLabels);

      const definitions = queryAllByRole('definition');

      expect(definitions.map(
        (definition) => definition.textContent)
      ).toEqual(expectedDefinitions);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <UserPage />,
        { store: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the user clicks the Change Password link', () => {
      it('should display the change password form', async () => {
        const { getByRole, getByText } = render(
          <UserPage />,
          { store: true },
        );
        const link = getByText('Change Password');

        await userEvent.click(link);

        const button = getByRole('button', { name: 'Update Password' });

        expect(button).toBeVisible();
      });

      it('should match the snapshot', async () => {
        const { asFragment, getByText } = render(
          <UserPage />,
          { store: true },
        );
        const link = getByText('Change Password');

        await userEvent.click(link);

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
