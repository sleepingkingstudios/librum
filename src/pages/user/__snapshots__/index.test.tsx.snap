// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserPage> when the user has failed to load should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <ul
      aria-label="navigation"
    >
      <li
        aria-label="navigation-item"
      >
        Nav Item @ /nav-item
      </li>
      <li>
        Dropdown
        <ul>
          <li
            aria-label="navigation-item"
          >
            Dropdown Item @ /dropdown-item
          </li>
        </ul>
      </li>
    </ul>
    <h1>
      Current User
    </h1>
    <div
      class="relative"
    >
      <dl
        class="data-list"
      >
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Name
          </dt>
          <dd
            class="data-list-value"
          >
            <span
              class="text-muted"
            >
              (none)
            </span>
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Email
          </dt>
          <dd
            class="data-list-value"
          >
            <span
              class="text-muted"
            >
              (none)
            </span>
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Role
          </dt>
          <dd
            class="data-list-value"
          >
            <span
              class="text-muted"
            >
              (none)
            </span>
          </dd>
        </div>
      </dl>
      <h2
        class="mb-2"
      >
        Security
      </h2>
      <a>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-user-lock pr-1"
          data-icon="user-lock"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H392.6c-5.4-9.4-8.6-20.3-8.6-32V352c0-2.1 .1-4.2 .3-6.3c-31-26-71-41.7-114.6-41.7H178.3zM528 240c17.7 0 32 14.3 32 32v48H496V272c0-17.7 14.3-32 32-32zm-80 32v48c-17.7 0-32 14.3-32 32V480c0 17.7 14.3 32 32 32H608c17.7 0 32-14.3 32-32V352c0-17.7-14.3-32-32-32V272c0-44.2-35.8-80-80-80s-80 35.8-80 80z"
            fill="currentColor"
          />
        </svg>
        Change Password
      </a>
    </div>
    <ul
      aria-label="breadcrumbs"
    >
      <li
        aria-label="breadcrumb"
      >
        Home @ /
      </li>
      <li
        aria-label="breadcrumb"
      >
        Current User @ /user
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<UserPage> when the user has successfully loaded should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <ul
      aria-label="navigation"
    >
      <li
        aria-label="navigation-item"
      >
        Nav Item @ /nav-item
      </li>
      <li>
        Dropdown
        <ul>
          <li
            aria-label="navigation-item"
          >
            Dropdown Item @ /dropdown-item
          </li>
        </ul>
      </li>
    </ul>
    <h1>
      Current User
    </h1>
    <div
      class="relative"
    >
      <dl
        class="data-list"
      >
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Name
          </dt>
          <dd
            class="data-list-value"
          >
            Alan Bradley
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Email
          </dt>
          <dd
            class="data-list-value"
          >
            alan.bradley@example.com
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Role
          </dt>
          <dd
            class="data-list-value"
          >
            User
          </dd>
        </div>
      </dl>
      <h2
        class="mb-2"
      >
        Security
      </h2>
      <a>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-user-lock pr-1"
          data-icon="user-lock"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H392.6c-5.4-9.4-8.6-20.3-8.6-32V352c0-2.1 .1-4.2 .3-6.3c-31-26-71-41.7-114.6-41.7H178.3zM528 240c17.7 0 32 14.3 32 32v48H496V272c0-17.7 14.3-32 32-32zm-80 32v48c-17.7 0-32 14.3-32 32V480c0 17.7 14.3 32 32 32H608c17.7 0 32-14.3 32-32V352c0-17.7-14.3-32-32-32V272c0-44.2-35.8-80-80-80s-80 35.8-80 80z"
            fill="currentColor"
          />
        </svg>
        Change Password
      </a>
    </div>
    <ul
      aria-label="breadcrumbs"
    >
      <li
        aria-label="breadcrumb"
      >
        Home @ /
      </li>
      <li
        aria-label="breadcrumb"
      >
        Current User @ /user
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<UserPage> when the user has successfully loaded when the user clicks the Change Password link should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <ul
      aria-label="navigation"
    >
      <li
        aria-label="navigation-item"
      >
        Nav Item @ /nav-item
      </li>
      <li>
        Dropdown
        <ul>
          <li
            aria-label="navigation-item"
          >
            Dropdown Item @ /dropdown-item
          </li>
        </ul>
      </li>
    </ul>
    <h1>
      Current User
    </h1>
    <div
      class="relative"
    >
      <dl
        class="data-list"
      >
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Name
          </dt>
          <dd
            class="data-list-value"
          >
            Alan Bradley
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Email
          </dt>
          <dd
            class="data-list-value"
          >
            alan.bradley@example.com
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Role
          </dt>
          <dd
            class="data-list-value"
          >
            User
          </dd>
        </div>
      </dl>
      <h2
        class="mb-2"
      >
        Security
      </h2>
      <form
        action="#"
        class="w-full lg:w-1/2 relative"
      >
        <div
          class=""
        >
          <div
            class="form-field"
          >
            <label
              class="form-label"
              for="old_password"
            >
              Old Password
            </label>
            <input
              class="form-input"
              id="old_password"
              name="oldPassword"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-field"
          >
            <label
              class="form-label"
              for="new_password"
            >
              New Password
            </label>
            <input
              class="form-input"
              id="new_password"
              name="newPassword"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-field"
          >
            <label
              class="form-label"
              for="confirm_password"
            >
              Confirm Password
            </label>
            <input
              class="form-input"
              id="confirm_password"
              name="confirmPassword"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="gap-3 sm:grid sm:grid-cols-1 lg:grid-cols-2"
        >
          <button
            class="form-button"
            type="button"
          >
            Cancel
          </button>
          <button
            class="form-button button-info"
            type="submit"
          >
            Update Password
          </button>
        </div>
      </form>
    </div>
    <ul
      aria-label="breadcrumbs"
    >
      <li
        aria-label="breadcrumb"
      >
        Home @ /
      </li>
      <li
        aria-label="breadcrumb"
      >
        Current User @ /user
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<UserPage> when the user is loading should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <ul
      aria-label="navigation"
    >
      <li
        aria-label="navigation-item"
      >
        Nav Item @ /nav-item
      </li>
      <li>
        Dropdown
        <ul>
          <li
            aria-label="navigation-item"
          >
            Dropdown Item @ /dropdown-item
          </li>
        </ul>
      </li>
    </ul>
    <h1>
      Current User
    </h1>
    <div
      class="relative"
    >
      <div
        class="form-loading-overlay absolute grid h-full place-content-center w-full"
      >
        <div
          class="inline-block text-center"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-user h-10 w-10"
            data-icon="user"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
              fill="currentColor"
            />
          </svg>
          <br />
          Loading User...
        </div>
      </div>
      <dl
        class="data-list"
      >
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Name
          </dt>
          <dd
            class="data-list-value"
          >
            <span
              class="text-muted"
            >
              (none)
            </span>
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Email
          </dt>
          <dd
            class="data-list-value"
          >
            <span
              class="text-muted"
            >
              (none)
            </span>
          </dd>
        </div>
        <div
          class="data-list-item"
        >
          <dt
            class="data-list-label"
          >
            Role
          </dt>
          <dd
            class="data-list-value"
          >
            <span
              class="text-muted"
            >
              (none)
            </span>
          </dd>
        </div>
      </dl>
      <h2
        class="mb-2"
      >
        Security
      </h2>
      Loading...
    </div>
    <ul
      aria-label="breadcrumbs"
    >
      <li
        aria-label="breadcrumb"
      >
        Home @ /
      </li>
      <li
        aria-label="breadcrumb"
      >
        Current User @ /user
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
